File: get_quotes.py
from urllib.request import urlopen
import datetime


def get_quotes(symbol, exchange, interval=86400, period="10Y"):
    """
    Downloads and returns historical quotes in csv

    symbol
    The code of the security. For example, GOOG for Google or EURUSD for the Euro/Dollar currency pair.

    exchange
    The exchange where the security is listed. For example, NASDAQ for GOOG or CURRENCY for EURUSD.

    interval
    A number in seconds. Its minimum value is 60 seconds.

    period
    The period of time from which data will be returned. Always returns the most recent data.
    Examples of this parameter are 1d (one day), 1w (one week), 1m (one month), or 1y (one year).
    """
    symbol = symbol.upper()
    exchange = exchange.upper()

    url = "https://www.google.com/finance/getprices?q=" +\
          symbol + "&x=" + exchange + "&i=" + str(interval) + "&p=" + str(period) +\
          '&f=d,c,v,k,o,h,l'

    data = urlopen(url).read().decode('utf-8')
    rows = data.splitlines()

    # text_file = open("output.txt", "w")
    # text_file.write(data)
    # text_file.close()

    # with open('output.txt', 'r') as text_file:
    #     rows = [line.rstrip('\n') for line in text_file]

    if len(rows) < 7:
        print('Incorrect parameters')
        return

    timezone_offset = 0
    absolute_date = 0

    result = "UTC_Date,Market_Date,Open,High,Low,Close,Volume\n"

    for i in range(6, len(rows)):
        if rows[i][:len("TIMEZONE_OFFSET=")] == "TIMEZONE_OFFSET=":
            timezone_offset = int(rows[i][len("TIMEZONE_OFFSET="):])
            continue

        temp = rows[i].split(",")

        if temp[0][:1] == "a":
            absolute_date = int(temp[0][1:])
            utc_date = absolute_date
        else:
            utc_date = absolute_date + int(temp[0]) * interval

        market_date = datetime.datetime.fromtimestamp(utc_date + timezone_offset * 60).strftime('%Y-%m-%d %H:%M:%S')

        open_col = temp[4]
        high_col = temp[2]
        low_col = temp[3]
        close_col = temp[1]
        volume_col = temp[5]

        result += str(utc_date) + "," + market_date + "," +\
                  open_col + "," + high_col + "," + low_col + "," + close_col + "," + volume_col + "\n"

    return result


# print(get_quotes("AAPL", "NASD", interval=3600, period="100d"))
# get_quotes("AAPL", "NASD")
# get_quotes("GBPUSD", "CURRENCY", interval=900, period="3d")
# help(get_quotes)

File: getQuotes.r
library(magrittr)

GOOG <- getQuotes("GOOG", "NASD", 1800, "5d")
GBPUSD <- getQuotes("GBPUSD", "CURRENCY", 900, "1d")
AAPL <- getQuotes("AAPL", "NYSE", 3600, "10d")

symbol <- "AAPL"
exchange <- "NASD"
interval <- 3600
period <- "10d"

getQuotes <- function(symbol, exchange, interval = 86400, period = "10Y") {
  url <- paste0("https://www.google.com/finance/getprices?q=",
                symbol,"&x=",exchange,"&i=",interval,"&p=",period)
  url_content <- readLines(url)
  
  j <- 1
  
  Date <- numeric()
  UTC <- numeric()
  Local <- numeric()
  Market <- numeric()
  Open <- numeric()
  High <- numeric()
  Low <- numeric()
  Close <- numeric()
  Volume <- numeric()
  
  if (length(url_content) > 7)
    for (i in 7:length(url_content)) {
      if (substr(url_content[i], 1, nchar("TIMEZONE_OFFSET=")) == "TIMEZONE_OFFSET=")
        timezone_offset <- url_content[i] %>%
          substring(nchar("TIMEZONE_OFFSET=")+1) %>% as.numeric
      else {
        temp <- url_content[i] %>% strsplit(",") %>% unlist
        
        if (substring(temp[1], 1, 1) == "a") {
          absolute_date <- temp[1] %>% substring(2) %>% as.numeric %>% as.POSIXct(origin="1970-01-01")
          Date[j] <- absolute_date
        }
        else {
          Date[j] <- absolute_date + as.numeric(temp[1]) * interval
        }
        
        UTC[j] <- Date[j] %>%
          as.POSIXct(origin="1970-01-01", tz="UTC") %>%
          format("%d.%m.%Y %H:%M:%OS")
        Market[j] <- (Date[j] + timezone_offset * 60) %>%
          as.POSIXct(origin="1970-01-01", tz="UTC") %>%
          format("%d.%m.%Y %H:%M:%OS")
        
        Open[j] <- temp[5] %>% as.numeric
        High[j] <- temp[3] %>% as.numeric
        Low[j] <- temp[4] %>% as.numeric
        Close[j] <- temp[2] %>% as.numeric
        Volume[j] <- temp[6] %>% as.numeric
        j <- j + 1
      }
    }
  else
    warning("Incorrect parameters")
  
  cat(symbol, "-", substring(url_content[1], nchar("EXCHANGE%3D")+1), "\n", length(Date), "obs.")
  
  return(data.frame(UTC, Market, Open, High, Low, Close, Volume, stringsAsFactors = FALSE))
}
